### PR TITLE
mockrelay: add DelayProfile latency model and in-repo sequence docs

### PR DESCRIPTION
**Pull request by philsphicas**
## Summary

mockrelay now reproduces the wall-clock cost of an Azure Relay
rendezvous through a per-lane `DelayProfile`, and the repo gains
canonical sequence diagrams documenting the wire-level behaviour
the model is calibrated against.

## `DelayProfile`

A `DelayProfile` carries five durations:

- `SLatency` — one-way propagation on the sender ↔ relay node leg.
- `LLatency` — one-way propagation on the listener ↔ relay node leg.
- `DNSLookup` — per-handler DNS cost (Go's resolver does not cache).
- `AuthInternal` — SAS-validation work on the relay node, paid in
  `handleListen` and `handleConnect`.
- `MatchMakeInternal` — matchmake/dispatch work on the relay node, paid
  in `handleConnect` while pairing with a registered listener.

Two presets are exported:

- `DelayProfileZero` — no synthetic delay anywhere (Go zero value,
  used by benchmarks and by the default `MockBackend{}`).
- `DelayProfileDefault` — calibrated to observed Azure warm timings
  (~755 ms per `ConnectLatency_Serial_PortForward` iter against a
  real namespace). The in-process mock with this preset delivers
  749-751 ms per iter, matching within ~5 ms.

The sleeps are placed surgically in `handleListen`, `handleConnect`,
and `handleAccept` to mirror the hop sequence verified against
wireshark captures of the real `*.servicebus.windows.net` rendezvous.
A `paired` channel synchronises the dual-101 emission so the sender
and listener pay their final response leg in parallel, matching the
observed back-to-back 101 frames.

## Pipelined bridge with byte budget

The bridge forwarding loop models pipelined wire propagation: each
queued WebSocket frame carries an `arriveBy` deadline of
`enqueue_time + S + L` that the writer goroutine honours before
delivering. Multiple frames are in flight concurrently per direction,
so:

- A single round-trip echo pays `2·(S+L)` (one S+L per direction).
- Streaming N back-to-back frames in one direction pays `~(S+L)`
  end-to-end (the bandwidth-delay product, not stop-and-wait
  serialisation).

A 12 MiB per-direction byte budget bounds in-flight bytes via a
weighted semaphore, calibrated to empirical Azure observations:
with a non-reading listener, real Azure backpressures the sender
at ~14-16 MiB regardless of frame size (probed at 1 KiB, 32 KiB,
and 256 KiB frames). Single oversized frames cap at the full budget
and serialise behind any frames already queued, matching Azure's
observed behaviour (no per-frame size limit; aggregate-bytes
backpressure only). The writer releases each frame's weight via
`defer`, so writer errors and ctx cancellations free the budget
and keep the reader's `Acquire` from hanging.

On source-side close, `pipelinedCopy` bounds the wait for the writer
to drain queued frames at `delay + 100 ms`. Without this bound, a
slow or dead downstream peer could hide the upstream close from
`bridgeWS` indefinitely.

## Sequence docs

`docs/internals/sequences/` adds two standalone wire-level sequence
documents:

- `port-forward.md` — port-forward wire sequence (5 phases) with
  per-leg lane annotations, a critical-path hop count table, and
  the key facts the `DelayProfile` model encodes.
- `socks5.md` — SOCKS5 wire sequence (5 phases) covering the local
  greeting/CONNECT, the envelope/target-dial/REP round-trip, and
  the resulting extra bridge round-trip on the critical path.

## Callers

`MockBackend.DelayProfile` (defaulting to `DelayProfileZero`) replaces
the removed `MockBackend.RendezvousDelay` knob and the removed
`mockrelay.WithAcceptDelay` server option. Pipeline-throughput
scenarios continue to use the zero profile; latency-sensitive e2e
tests now pass `server.DelayProfileDefault` explicitly.

## Validation

- Mockrelay unit tests, including new `DelayProfile` and byte-budget
  tests, pass under `-race -count=1`. The byte-budget tests target
  internal `runPipelinedReader`/`runPipelinedWriter` so assertions
  fail deterministically if the semaphore is removed.
- `TestE2E_Mock` passes under `-race -tags=e2e -timeout=900s`
  (the wall-clock increase from earlier runs is the lane-modelled
  bridge faithfully paying `S+L` per direction).
- All four modules build clean and `golangci-lint` is clean.
- All mermaid blocks render under `mmdc`; `prettier --check .` is
  clean.

### CODE REVIEW

File: mockrelay/server/delay_profile.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+// WithDelayProfile arms the per-lane synthetic-delay model. The zero
+// profile applies no delay anywhere — pass DelayProfileDefault (or
+// build your own) for fidelity. Only effective when set via
+// NewServerForTesting; the production constructor never installs a
+// profile, so a Server returned from NewServer is always profile-less.
+func WithDelayProfile(p DelayProfile) Option {
+	return func(s *Server) error {
+		s.delayProfile = p
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply mockrelay/server/delay_profile.go:122]: Good catch — DelayProfile now has a validate() method that rejects any negative duration (per-field, with the offending field name in the error), and WithDelayProfile calls it and wraps the error before returning. Added table-driven unit tests covering all five fields plus a positive test that zero values are accepted.

File: mockrelay/server/bridge_ws.go [hunk 2]
```diff
@@ -4,65 +4,236 @@ import (
...
+	weight   int64 // semaphore weight acquired for this frame
+}
+
+// bridgeFrameWeight returns the semaphore weight for a frame of n
+// bytes, clamping to bridgeByteBudget. A single oversized frame
+// acquires the entire budget and serialises behind any frames
+// already queued; this matches real Azure, which has no observed
+// per-frame size limit (probed up to 32 MiB) but caps aggregate
+// in-flight bytes. Reader and writer share this helper so the
+// weight acquired and the weight released are guaranteed equal.
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply mockrelay/server/bridge_ws.go]: The comment now spells out that the 16 MiB ceiling is enforced by bridgeReadLimit at the WS read layer (control.go), distinct from the per-direction in-flight byte budget that bridgeFrameWeight feeds the semaphore.

File: docs/internals/sequences/socks5.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
+# SOCKS5 rendezvous
+
+Wire-level sequence for a SOCKS5 rendezvous from cold state:
+
+1. The Listener opens a control WebSocket to the Relay
+   (`action=listen`) and waits.
+2. A Client opens a TCP connection to the local SOCKS5 server
+   (in-process on the Sender).
+3. Client and Sender complete the SOCKS5 greeting and CONNECT.
+4. The Sender rendezvous'es with the Relay.
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply docs/internals/sequences/socks5.md]: Fixed — reads "performs the rendezvous with the Relay" now.

File: docs/internals/sequences/socks5.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply docs/internals/sequences/socks5.md]: Fixed — reads "performs the rendezvous with the Relay" now.

File: mockrelay/server/bridge_ws.go [hunk 5]
```diff
@@ -4,65 +4,236 @@ import (
...
+	// freshest queued frame's arriveBy. If the writer is parked on
+	// a slow or dead destination peer, returning readErr promptly
+	// lets bridgeWS cancel the other direction and close `dst`,
+	// which unblocks the writer; the goroutine completes in the
+	// background. Without this bound, a slow downstream peer would
+	// hide the upstream peer's close from bridgeWS indefinitely.
+	drainBudget := delay + 100*time.Millisecond
+	select {
+	case writeErr := <-writerDone:
+		if readErr != nil {
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply mockrelay/server/bridge_ws.go:177]: Good catch — pipelinedCopy now prefers the writer error when the read error is the cancellation we induced ourselves. The logic checks errors.Is(readErr, context.Canceled) AND ctx.Err() == nil (parent ctx still alive proves the cancel came from our writer-induced copyCtx cancel, not from a parent shutdown). Added a regression test (TestPipelinedCopy_PrefersWriterErrorOverInducedCancel) that asserts the writer's sentinel error wins when the reader's read fails with context.Canceled because writerDone fired first.

File: mockrelay/server/rendezvous.go [hunk 6]
```diff
@@ -136,41 +160,80 @@ func (s *Server) handleConnect(w http.ResponseWriter, r *http.Request, entity st
...
+			http.Error(w, "rendezvous timeout", http.StatusGatewayTimeout)
+			return
+		}
+		return // client gone
+	case <-pending.paired:
+	}
+
+	// 101 transit leg to sender — runs in parallel with handleAccept's
+	// own response leg (the listener-side 101).
+	if !sleepContext(ctx, hopsResponse*p.SLatency) {
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply mockrelay/server/rendezvous.go]: Good catch — the post-pair sleep now uses waitCtx, so a stall during the 101 transit honours RendezvousTimeout. If waitCtx fires with DeadlineExceeded AND the parent ctx is still alive, the handler logs a warning and emits 504. The parent-ctx guard avoids misclassifying a parent-deadline cancellation as a rendezvous timeout. Added TestRendezvousTimeoutOn101Transit to lock this in.

File: e2e/backends/mock/e2e_test.go [hunk 7]
```diff
@@ -7,14 +7,20 @@ import (
...
 // TestE2E_Mock runs the shared e2e scenarios against the in-process
 // MockBackend. This is the always-on side of the mock-vs-Azure
 // conformance matrix; it runs in `make e2e-mock` and pairs with
 // TestE2E_Azure in e2e/backends/azure to surface any behavioural
 // divergence between the mock and Azure.
+//
+// DelayProfileDefault is the recommended profile for e2e-style runs:
+// it approximates the wireshark-observed wall-clock shape from the
+// AUE captures so timing thresholds calibrated against Azure also
+// fire against the mock.
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply e2e/backends/mock/e2e_test.go]: Good catch — comment now reads "real Azure Relay captures" instead of the internal lab shorthand.

File: mockrelay/server/rendezvous.go [hunk 8]
```diff
@@ -136,41 +160,89 @@ func (s *Server) handleConnect(w http.ResponseWriter, r *http.Request, entity st
...
+	if !sleepContext(waitCtx, hopsResponse*p.SLatency) {
+		if ctx.Err() == nil && errors.Is(waitCtx.Err(), context.DeadlineExceeded) {
+			s.log.Warn("rendezvous timeout (101 transit)", "entity", entity, "id", id)
+			http.Error(w, "rendezvous timeout", http.StatusGatewayTimeout)
+		}
+		return
+	}
+	// Pre-Accept defensive re-check: if pending.ready closed during
+	// the 1S sleep AND listenerWS is nil, the listener's WS Accept
+	// failed on its side. Bail with 503 pre-upgrade rather than
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply mockrelay/server/rendezvous.go]: Good catch — reworded to "the 101-transit sleep above" so it stays accurate regardless of hopsResponse or profile values.

File: docs/internals/sequences/port-forward.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
+# Port-forward rendezvous
+
+Wire-level sequence for a single port-forward rendezvous from cold
+state:
+
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply docs/internals/sequences/port-forward.md:5]: Good catch — the README was intentionally dropped during earlier iteration (each doc stands alone with its own conventions preamble), but the PR description and commit message hadn't caught up. Rewrote the Sequence docs section in both to describe what's actually being introduced (port-forward.md + socks5.md as two standalone wire-level documents).

File: mockrelay/server/bridge_ws.go [hunk 10]
```diff
@@ -4,65 +4,249 @@ import (
...
+		if readErr != nil && errors.Is(readErr, context.Canceled) && ctx.Err() == nil && writeErr != nil {
+			return writeErr
+		}
+		if readErr != nil {
+			return readErr
+		}
+		return writeErr
+	case <-time.After(drainBudget):
+		return readErr
+	}
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply mockrelay/server/bridge_ws.go:183]: Good catch — switched to time.NewTimer with defer Stop() so the timer is cleaned up immediately when writerDone wins, instead of being left in the runtime's heap until drainBudget fires.
